### PR TITLE
feat: support `unnest` in GROUP BY clause

### PR DESCRIPTION
**Pull request by JasonLi-cn**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/11470

## Rationale for this change

support SQLs like:
```
SELECT unnest(list_type_col) c0, count(1) FROM tab GROUP BY c0;
SELECT unnest(unnest(recursive_list_type_col)) c0, count(1) FROM tab GROUP BY c0;
```

## Are these changes tested?

Yes. Has added some test cases into `unnest.slt`.

## Are there any user-facing changes?

Simplified SQL syntax, eg:

```
SELECT t0.c0, count(1) FROM (SELECT unnest(list_col) c0 FROM tal) t0 GROUP BY t0.c0;
```
can replace by
```
SELECT unnest(list_col) c0, count(1) FROM tal GROUP BY c0;
```

### CODE REVIEW

**Comment by jayzhan211** [comment]
I wonder why we need to specialize group by expression for unnest, does that mean we need to specialize unnest with filter by too? 🤔

**Comment by 2010YOUY01** [comment]
Thank you for the new feature, I have some thoughts for an alternative design:

Recursive aggregate functions are generally unsupported: e.g. `select sum(sum(v1)) from t1` is not supported in DataFusion/Postgres/DuckDB.
For this `unnest()` case, a recursive call is required, DuckDB uses: https://duckdb.org/docs/sql/query_syntax/unnest.html#recursive-unnest

What do you think about implementing DuckDB syntax, I think it's better for:
1. Keep the SQL syntax consistent and avoid nested aggregate function
2. Avoid changing planner code (which might be error-prone), and keep the unnest logic within the aggregate function body

**Comment by JasonLi-cn** [comment]
> I wonder why we need to specialize group by expression for unnest, does that mean we need to specialize unnest with filter by too? 🤔

- Pg supports group by `unnest` expression;
- Our user needs this syntax support. 

So, I tried to implement it in the community version.

As far as I know, pg does not currently support unnest in filter, so I'm not sure it needs to be implemented.

**Comment by JasonLi-cn** [comment]
> Thank you for the new feature, I have some thoughts for an alternative design:
> 
> Recursive aggregate functions are generally unsupported: e.g. `select sum(sum(v1)) from t1` is not supported in DataFusion/Postgres/DuckDB. For this `unnest()` case, a recursive call is required, DuckDB uses: https://duckdb.org/docs/sql/query_syntax/unnest.html#recursive-unnest
> 
> What do you think about implementing DuckDB syntax, I think it's better for:
> 
> 1. Keep the SQL syntax consistent and avoid nested aggregate function
> 2. Avoid changing planner code (which might be error-prone), and keep the recursive unnest logic within the aggregate function body

Thank you for your reply.

1 If "keep the recursive unnest logic within the aggregate function body" is implemented, at least the following work needs to be done:
(1) Improve the logic of `Expr::Unnest` processing in `TypeCoercionRewriter`;
(2) Implement `create_physical_expr` function to support `Expr::Unnest`;
(3) Maybe there are other tasks...
Also, I would like to check whether you are talking about `aggregate function` or `AggregateExec` here. Because pg currently does not support aggregation functions that include `unnest`. (Of course, it is not appropriate to illustrate this issue only by pg here)

2 If 'Recursive Unnest' is handled like DuckDB, how do we implement the unnest(unnest(list_column)) syntax when the user needs it? By optimization rules?

**Comment by 2010YOUY01** [comment]
> > Thank you for the new feature, I have some thoughts for an alternative design:
> > Recursive aggregate functions are generally unsupported: e.g. `select sum(sum(v1)) from t1` is not supported in DataFusion/Postgres/DuckDB. For this `unnest()` case, a recursive call is required, DuckDB uses: https://duckdb.org/docs/sql/query_syntax/unnest.html#recursive-unnest
> > What do you think about implementing DuckDB syntax, I think it's better for:
> > 
> > 1. Keep the SQL syntax consistent and avoid nested aggregate function
> > 2. Avoid changing planner code (which might be error-prone), and keep the recursive unnest logic within the aggregate function body
> 
> Thank you for your reply.
> 
> 1 If "keep the recursive unnest logic within the aggregate function body" is implemented, at least the following work needs to be done: (1) Improve the logic of `Expr::Unnest` processing in `TypeCoercionRewriter`; (2) Implement `create_physical_expr` function to support `Expr::Unnest`; (3) Maybe there are other tasks... Also, I would like to check whether you are talking about `aggregate function` or `AggregateExec` here. Because pg currently does not support aggregation functions that include `unnest`. (Of course, it is not appropriate to illustrate this issue only by pg here)
> 
> 2 If 'Recursive Unnest' is handled like DuckDB, how do we implement the unnest(unnest(list_column)) syntax when the user needs it? By optimization rules?

@JasonLi-cn I mistakenly thought `unnest` was some aggregate function... please ignore my previous reply, and also sorry for the confusion.

**Comment by jayzhan211** [comment]
@JasonLi-cn Is it better to move the logic into `try_process_unnest`?

File: datafusion/sql/src/select.rs [hunk 1]
```diff
@@ -354,6 +358,102 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        &self,
+        agg: &Aggregate,
+    ) -> Result<(LogicalPlan, Vec<Expr>)> {
+        let mut aggr_expr_using_columns: Option<HashSet<Expr>> = None;
+
+        let input = agg.input.as_ref();
+        let group_by_exprs = &agg.group_expr;
+        let aggr_exprs = &agg.aggr_expr;
+
+        // rewrite group_by_exprs
```
Comment (alamb) [inline datafusion/sql/src/select.rs:395]: I wonder if you could add some documentation comments to this function that explains what it does

Ideally such documentation could include an example showing what it is necessary to handle group by specially

File: datafusion/sql/src/select.rs [hunk 2]
```diff
@@ -354,6 +358,102 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                            // As the closure always returns Ok, this "can't" error
+                            .expect("Unexpected error");
+                        }
+                        aggr_expr_using_columns = Some(columns.clone());
+                        columns
+                    }
+                };
+                projection_exprs.extend(inner_projection_exprs);
+
+                let plan = LogicalPlanBuilder::from(intermediate_plan.clone())
```
Comment (alamb) [inline datafusion/sql/src/select.rs:444]: maybe you could avoid the clone with something like

```suggestion
                let intermediate_plan = LogicalPlanBuilder::from(intermediate_plan,
```

File: datafusion/sqllogictest/test_files/unnest.slt [hunk 3]
```diff
@@ -556,4 +554,126 @@ physical_plan
...
+1
+2
+3
+4
+5
+6
+12
+
+query II
+select unnest(column1) c1, unnest(column2) c2 from unnest_table group by c1, c2 order by c1, c2;
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/unnest.slt:575]: Could you also please add a test with grouping directly in the group by list (not a reference to the select list)

Like

```suggestion
select unnest(column1), count(*) from unnest_table group by unnest(column1);
```

**Review by alamb — APPROVED** [review]
Thank you @JasonLi-cn  -- I think this PR looks good to me. Thank you for the contribution

I think adding some comments about what `try_process_group_by_unnest` is doing and why would help improve this code but I don't think it is required prior to merge

c @jonahgao

**Comment by JasonLi-cn** [comment]
> Thank you @JasonLi-cn -- I think this PR looks good to me. Thank you for the contribution
> 
> I think adding some comments about what `try_process_group_by_unnest` is doing and why would help improve this code but I don't think it is required prior to merge
> 
> c @jonahgao

Thank you @alamb for your review. I have completed the code modification according to your modification suggestions.

File: datafusion/sql/src/select.rs [hunk 4]
```diff
@@ -354,6 +358,118 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        //
+        // will be transformed into
+        //
+        // ```
+        // Aggregate: groupBy=[[unnest(tab.array_col)]], aggr=[[]]
+        //   Unnest: lists[unnest(tab.array_col)] structs[]
+        //     Projection: tab.array_col AS unnest(tab.array_col)
+        //       TableScan: tab
+        // ```
+        let mut intermediate_plan = input.clone();
```
Comment (jonahgao) [inline datafusion/sql/src/select.rs:413]: Is it possible to avoid this clone?

File: datafusion/sql/src/select.rs [hunk 5]
```diff
@@ -354,6 +358,118 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                }
+                _ => Ok(input),
+            },
+            _ => Ok(input),
+        }
+    }
+
+    /// Try converting Unnest(Expr) of group by to Unnest/Projection
+    /// Return the new input and group_by_exprs of Aggregate.
+    fn try_process_group_by_unnest(
```
Comment (jonahgao) [inline datafusion/sql/src/select.rs:387]: This function seems to have some code repetition with function [try_process_unnest](https://github.com/apache/datafusion/blob/f204869ff55bb3e39cf23fc0a34ebd5021e6773f/datafusion/sql/src/select.rs#L295). 

I wonder if there's a better way to handle this, such as planning unnest before aggregation, and then reusing the current group-by planning logic. This seems more intuitive to me. But I'm not sure about it.

File: datafusion/sql/src/select.rs [hunk 6]
```diff
@@ -297,6 +298,9 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         input: LogicalPlan,
         select_exprs: Vec<Expr>,
     ) -> Result<LogicalPlan> {
+        // Try process group by unnest
+        let input = self.try_process_aggregate_unnest(input)?;
```
Comment (jonahgao) [inline datafusion/sql/src/select.rs:302]: If unnest has already been processed by `try_process_aggregate_unnest`, does the following logic for handling `unnest` become redundant?

**Review by jonahgao — APPROVED** [review]
I left some comments, but I think it can be merged as is. It works well. Perhaps we can try to improve it through subsequent PRs.

File: datafusion/sql/src/select.rs (hunk 6 above)
Comment (JasonLi-cn) [reply datafusion/sql/src/select.rs:302]: I agree with you. We need to verify this logic in the future

File: datafusion/sql/src/select.rs (hunk 4 above)
Comment (alamb) [reply datafusion/sql/src/select.rs]: I pushed 810ca8b80 which avoids the cloning in this function

File: datafusion/sql/src/select.rs (hunk 6 above)
Comment (alamb) [reply datafusion/sql/src/select.rs:303]: Tracked in https://github.com/apache/datafusion/issues/11498

File: datafusion/sql/src/select.rs (hunk 5 above)
Comment (alamb) [reply datafusion/sql/src/select.rs:384]: Tracked in https://github.com/apache/datafusion/issues/11498

**Comment by alamb** [comment]
I merged up to resolve a conflict as well

**Comment by JasonLi-cn** [comment]
> I merged up to resolve a conflict as well

Thank you @alamb for your help to modify the code, I learned a lot from it.

**Comment by jonahgao** [comment]
Thanks @JasonLi-cn @alamb

**Pull request merged**